packages/core/src/jsx-runtime.js

```diff
@@ -0,0 +1,174 @@
+  className += `${cache.key}-${serialized.name}`
+
+  const newProps = {}
+  for (let key in props) {
+    if (
+      hasOwnProperty.call(props, key) &&
+      key !== 'css' &&
+      key !== typePropName &&
+      (process.env.NODE_ENV === 'production' || key !== labelPropName)
+    ) {
+      newProps[key] = props[key]
+    }
+  }
+  newProps.ref = ref
+  newProps.className = className
+
+  const ele = jsx(type, newProps)
+  if (!isBrowser && rules !== undefined) {
+    let serializedNames = serialized.name
+    let next = serialized.next
+    while (next !== undefined) {
+      serializedNames += ' ' + next.name
+      next = next.next
+    }
+    return (
+      <React.Fragment>
+        <style
+          {...{
+            [`data-emotion-${cache.key}`]: serializedNames,
+            dangerouslySetInnerHTML: { __html: rules },
+            nonce: cache.sheet.nonce
+          }}
+        />
+        {ele}
+      </React.Fragment>
+    )
+  }
+  return ele
+}
+
+let Emotion = /* #__PURE__ */ withEmotionCache((props, cache, ref) => {
+  // use Context.read for the theme when it's stable
+  if (typeof props.css === 'function') {
+    return (
+      <ThemeContext.Consumer>
+        {theme => render(cache, props, theme, ref)}
+      </ThemeContext.Consumer>
+    )
+  }
+  return render(cache, props, null, ref)
+})
+
+if (process.env.NODE_ENV !== 'production') {
+  Emotion.displayName = 'EmotionCssPropInternal'
+}
+
+const jsx = (type, props, maybeKey) => {
+  let args = arguments
+
+  if (props == null || !hasOwnProperty.call(props, 'css')) {
+    return reactJsx(type, props, maybeKey)
+  }
+
+  if (
+    process.env.NODE_ENV !== 'production' &&
+    typeof props.css === 'string' &&
+    // check if there is a css declaration
+    props.css.indexOf(':') !== -1
+  ) {
+    throw new Error(
+      `Strings are not allowed as css prop values, please wrap it in a css template literal from '@emotion/css' like this: css\`${
+        props.css
+      }\``
+    )
+  }
+
+  let argsLength = args.length
+
+  let createElementArgArray = new Array(argsLength)
+
+  createElementArgArray[0] = Emotion
+  let newProps = {}
+
+  for (let key in props) {
+    if (hasOwnProperty.call(props, key)) {
+      newProps[key] = props[key]
+    }
+  }
+  newProps[typePropName] = type
+  if (process.env.NODE_ENV !== 'production') {
+    let error = new Error()
+    if (error.stack) {
+      // chrome
+      let match = error.stack.match(
+        /at (?:Object\.|Module\.|)jsx.*\n\s+at ([A-Z][A-Za-z$]+) /
+      )
+      if (!match) {
+        // safari and firefox
+        match = error.stack.match(/.*\n([A-Z][A-Za-z$]+)@/)
+      }
+      if (match) {
+        newProps[labelPropName] = sanitizeIdentifier(match[1])
+      }
+    }
+  }
+
+  createElementArgArray[1] = newProps
+
+  for (let i = 2; i < argsLength; i++) {
+    createElementArgArray[i] = args[i]
+  }
+  // $FlowFixMe
+  return reactJsx.apply(null, createElementArgArray)
+}
+
+const jsxs = jsx
```

Comment:
https://github.com/facebook/react/pull/18299/files#diff-97fbb0b56908c06793243318ab1cb29aR19

![image](https://user-images.githubusercontent.com/4138357/90041438-e7c2ee00-dc8e-11ea-80d4-15c4e401c41e.png)

Reply:
I don't understand this part.

Looking at the initial RFC for create element changes 

> My proposal is that we compile <div>{a}{b}</div> to jsxs('div', {children: [a, b]}) and <div>{a}</div> to jsx('div', {children:a}). The jsxs function indicates that the top array was created by React. The nice property of this strategy is that even if you don't have separate build steps for PROD and DEV, we can still issue the key warning and still pay no cost in PROD.

and the line of code you linked

https://github.com/facebook/react/pull/18299/files#diff-97fbb0b56908c06793243318ab1cb29aR19

```jsx
const jsx = __DEV__ ? jsxWithValidationDynamic : jsxProd;
// we may want to special case jsxs internally to take advantage of static children.
// for now we can ship identical prod functions
const jsxs = __DEV__ ? jsxWithValidationStatic : jsxProd;
const jsxDEV = __DEV__ ? jsxWithValidation : undefined;
```

it seems to me that the difference between _jsxWithValidationStatic_ and _jsxWithValidationDynamic_ is signifcant during development, and Emotion shouldn't discard jsxs.

Reply:
If anything we should just forward to whatever React exports - but I'm quite hesitant about moving forward with this PR at this point in time as React 17 is not released yet. We'll have to wait for it and then think how to support both React 16 and React 17 at the same time or if we should cut a new major for React 17.

Reply:
> how to support both React 16 and React 17 at the same time

This would be preferable. I was undder impression that React 17 still supports React.createElement, and the only addition is `react/jsx-runtime` entrypoint.

Emotion would have to export three things to match React's exports.

1. current __jsx__ function to support old runtime, using __React.createElement__ underneath
   ```(type, props, ...children) => ReactElement```
2. a new __jsx__ function, using _jsx_ from `react/jsx-runtime` (*)
   ```(type, config, maybeKey, source?, self?) => ReactElement```
3. __jsxs__ function, using _jsxs__ from `react/jsx-runtime`

This PR already creates the the `@emotion/core/jsx-runtime` entrypoint.

(*) There's a warning about explicitly using new `jsx` functions in React docs.
It's probably meant for projects that don't implement their own `jsx` function, though.

> The functions inside react/jsx-runtime and react/jsx-dev-runtime must only be used by the compiler transform. If you need to manually create elements in your code, you should keep using React.createElement. It will continue to work and is not going away.
>
> _https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html_

I'm not sure yet how `react/jsx-dev-runtime` fits the puzzle.

Reply:
> This would be preferable. I was under impression that React 17 still supports React.createElement, and the only addition is react/jsx-runtime entrypoint.

Yes, that's correct.

> Emotion would have to export three things to match React's exports.

That's correct as well - the only problematic part is that TypeScript doesn't import fully automatic runtime with custom modules 😞 but that's kinda their limitation and not ours, even though it affects us greatly.

> (*) There's a warning about explicitly using new jsx functions in React docs.
> It's probably meant for projects that don't implement their own jsx function, though.

Yes, I believe we don't have to care about this warning - API of those functions have to be stable after all, otherwise, compilers would break, so we should be free to use them directly.

> I'm not sure yet how react/jsx-dev-runtime fits the puzzle.

I think it's the same - we should just forward stuff to the appropriate function and that's all.